format/ComputeIR.fbs

```diff
@@ -0,0 +1,521 @@
+  name:string (required);
+
+  /// The schema of the data source. This may be a partial schema (ignoring
+  /// unused fields), but it at least asserts the fields and types that are
+  /// expected to exist in the data source.
+  schema:Schema (required);
+
+  /// Optional opaque table serialization data, for passing engine-specific
+  /// instructions to enable the data to be accessed.
+  serde_type:string;
+  serde_data:[ubyte];
+}
+
+/// An auxiliary helper instruction to "include all columns" in the projection,
+/// sparing the IR producer the need to enumerate all column references in a
+/// projection.
+table StarSelection {
+  ref:TableReference;
+}
+
+/// A helper union to permit the "SELECT *, $expr0, ..." construct from SQL.
+union ProjectionExpr {
+  ArrayExpr,
+  StarSelection
+}
+
+/// Computes a new table given a set of column selections or array expressions.
+table Projection {
+  input:TableExpr (required);
+
+  /// Each expression must reference fields foudn in the input table
+  /// expression.
+  exprs:[ProjectionExpr] (required);
+}
+
+/// Select rows from table for given boolean condition.
+table Filter {
+  input:TableExpr (required);
+
+  /// Array expression using input table expression yielding boolean output
+  /// type.
+  condition:ArrayExpr (required);
+}
+
+/// A "group by" table aggregation: data is grouped using the group
+/// expressions, and the aggregate expressions are evaluated within each group.
+table Aggregate {
+  input:TableExpr;
+  aggregate_exprs:[ArrayExpr] (required);
+
+  /// Expressions to use as group keys. If not provided, then the aggregate
+  /// operation yields a table with a single row.
+  group_exprs:[ArrayExpr];
+
+  /// A filter condition for the aggregation which may include aggregate
+  /// functions.
+  having:[ArrayExpr];
+}
+
+/// Select up to the indicated number of rows from the input expression based
+/// on the first-emitted rows when evaluating the input expression. Generally,
+/// no particular order is guaranteed unless combined with a sort expression.
+table Limit {
+  input:TableExpr;
+
+  /// Number of logical rows to select from input.
+  limit:long;
+}
+
+enum RelationalJoinType : int {
+  INNER = 0,
+  LEFT = 1,
+  RIGHT = 2,
+  OUTER = 3,
+  SEMI = 4,
+  ANTI = 5,
+}
+
+/// A standard relational / SQL-style equijoin.
+///
+/// Providing no left/right columns produces the cross product of the two
+/// tables.
+table EqualityJoin {
+  // TODO: complete and document
+  type:RelationalJoinType = INNER;
+
+  left:TableExpr (required);
+  right:TableExpr (required);
+
+  /// The columns from the left expression to use when joining
+  left_columns:[ColumnReference] (required);
+
+  /// The columns from the right expression to use when joining. Must have the
+  /// same length as left_columns.
+  ///
+  /// If omitted, the names provided in left_columns must match the names in
+  /// the right expression.
+  right_columns:[ColumnReference];
+}
```

Comment:
SQL style joins allow any arbitrary prediate (often called `on_exprs` as they appear in the `ON` clause). The break out of left/right columns for equijoin is almost always required for performance reasons, but there can be other predicates). 

For example here is a query you can not represent using the Compute IR in this PR:

```sql
SELECT * 
FROM 
  orders LEFT JOIN lineitem ON (l_orderkey = o_orderkey AND l_comments LIKE '%one star%'`
```
 which would produce a values for all orders, even if they didn't have any "one star" reviews

Reply:
There is the `NonEqualityJoin` which allows for arbitrary expressions. This could be collapsed to be just a single expression-based join and leave it to the engine to decide how to execute the join

Reply:
I think having a single join construct for now makes the most sense. I don't see a reason to differentiate in the IR, since concerns about types of joins are IMO out of scope for the IR and I think in every case would fall solely onto the consumer to differentiate.